### ISSUE TITLE
`cabal v1-test` and `cabal new-test` do not recompile type class instances if default implementation changes

### ISSUE THREAD

**Issue by yeputons** [haskell/cabal#6045]
## Preparation
I have a `class MyClass` in `src/MyClass.hs` with a function `foo` with a default implementation. I also have an `instance MyClass Int` in `src/MyInstance.hs` which does not override `foo`, hence uses the default implementation from `src/MyClass.hs`. When I run `cabal test` (`v1-` or `new-`, does not matter), it compiles and runs successfully.

## Expected behaviour
If I update the default implementation of `foo` in `src/MyClass.hs` and run `cabal test`, it will recompile all files necessary so that change is reflected in newly compiled binary.

## Real behaviour
If I update the default implementation of `foo` in `src/MyClass.hs` and run `cabal test`, it will recompile `src/MyClass.hs`, but not `src/MyInstance.hs`, and the resulting binary will still use the old implementation. If I run `cabal clean` before `cabal test`, it works.

## Versions
```
cabal-install version 2.4.1.0
compiled using version 2.4.1.0 of the Cabal library 
The Glorious Glasgow Haskell Compilation System, version 8.6.3
```

I'm using Windows 7 x64 and have installed GHC using `ghcup`.

## Reproduction

[Example repo with reproduction instructions](https://github.com/yeputons/cabal-type-class-default-impl-rebuild)

## Extra

Please let me know if any extra information or actions would be helpful from my side.

I'd also appreciate any workarounds, e.g. forcing cabal to recompile a specific file without mass `cabal clean`.

**Comment by phadej** [comment]
Cannot reproduce with

```
[polinukli] /code/mess/cabal-type-class-default-impl-rebuild master % cabal --version
cabal-install version 3.0.0.0
compiled using version 3.0.0.0 of the Cabal library 
[polinukli] /code/mess/cabal-type-class-default-impl-rebuild master % dpkg -l|grep cabal
ii  cabal-install-3.0                          3.0+git20190421.1.38969a0-6~18.04            amd64        Command-line interface for Cabal and Hackage
```


```
[polinukli] /code/mess/cabal-type-class-default-impl-rebuild master % cat src/MyClass.hs  
module MyClass where
class MyClass a where
    foo :: a -> Int
    foo _ = undefined
    --foo _ = 10
[polinukli] /code/mess/cabal-type-class-default-impl-rebuild master % cabal v2-test 
Resolving dependencies...
Build profile: -w ghc-8.6.5 -O1
In order, the following will be built (use -v for more details):
 - demo-0.1.0.0 (lib) (first run)
 - demo-0.1.0.0 (test:map) (first run)
Configuring library for demo-0.1.0.0..
Preprocessing library for demo-0.1.0.0..
Building library for demo-0.1.0.0..
[1 of 2] Compiling MyClass          ( src/MyClass.hs, /code/mess/cabal-type-class-default-impl-rebuild/dist-newstyle/build/x86_64-linux/ghc-8.6.5/demo-0.1.0.0/build/MyClass.o )
[2 of 2] Compiling MyInstance       ( src/MyInstance.hs, /code/mess/cabal-type-class-default-impl-rebuild/dist-newstyle/build/x86_64-linux/ghc-8.6.5/demo-0.1.0.0/build/MyInstance.o )
Configuring test suite 'map' for demo-0.1.0.0..
Warning: The package has an extraneous version range for a dependency on an
internal library: demo -any && ==0.1.0.0. This version range includes the
current package but isn't needed as the current package's library will always
be used.
Preprocessing test suite 'map' for demo-0.1.0.0..
Building test suite 'map' for demo-0.1.0.0..
[1 of 1] Compiling Main             ( test/MyTest.hs, /code/mess/cabal-type-class-default-impl-rebuild/dist-newstyle/build/x86_64-linux/ghc-8.6.5/demo-0.1.0.0/t/map/build/map/map-tmp/Main.o )
Linking /code/mess/cabal-type-class-default-impl-rebuild/dist-newstyle/build/x86_64-linux/ghc-8.6.5/demo-0.1.0.0/t/map/build/map/map ...
Running 1 test suites...
Test suite map: RUNNING...
map: Prelude.undefined
CallStack (from HasCallStack):
  error, called at libraries/base/GHC/Err.hs:78:14 in base:GHC.Err
  undefined, called at src/MyClass.hs:4:13 in demo-0.1.0.0-inplace:MyClass
Test suite map: FAIL
Test suite logged to:
/code/mess/cabal-type-class-default-impl-rebuild/dist-newstyle/build/x86_64-linux/ghc-8.6.5/demo-0.1.0.0/t/map/test/demo-0.1.0.0-map.log
0 of 1 test suites (0 of 1 test cases) passed.
cabal: Tests failed for test:map from demo-0.1.0.0.

[polinukli] /code/mess/cabal-type-class-default-impl-rebuild master % vim src/MyClass.hs 
[polinukli] /code/mess/cabal-type-class-default-impl-rebuild master % cat src/MyClass.hs
module MyClass where
class MyClass a where
    foo :: a -> Int
    --foo _ = undefined
    foo _ = 10
[polinukli] /code/mess/cabal-type-class-default-impl-rebuild master % cabal v2-test     
Build profile: -w ghc-8.6.5 -O1
In order, the following will be built (use -v for more details):
 - demo-0.1.0.0 (lib) (file src/MyClass.hs changed)
 - demo-0.1.0.0 (test:map) (dependency rebuilt)
Preprocessing library for demo-0.1.0.0..
Building library for demo-0.1.0.0..
[1 of 2] Compiling MyClass          ( src/MyClass.hs, /code/mess/cabal-type-class-default-impl-rebuild/dist-newstyle/build/x86_64-linux/ghc-8.6.5/demo-0.1.0.0/build/MyClass.o )
[2 of 2] Compiling MyInstance       ( src/MyInstance.hs, /code/mess/cabal-type-class-default-impl-rebuild/dist-newstyle/build/x86_64-linux/ghc-8.6.5/demo-0.1.0.0/build/MyInstance.o ) [MyClass changed]
Preprocessing test suite 'map' for demo-0.1.0.0..
Building test suite 'map' for demo-0.1.0.0..
[1 of 1] Compiling Main             ( test/MyTest.hs, /code/mess/cabal-type-class-default-impl-rebuild/dist-newstyle/build/x86_64-linux/ghc-8.6.5/demo-0.1.0.0/t/map/build/map/map-tmp/Main.o ) [MyClass changed]
Linking /code/mess/cabal-type-class-default-impl-rebuild/dist-newstyle/build/x86_64-linux/ghc-8.6.5/demo-0.1.0.0/t/map/build/map/map ...
Running 1 test suites...
Test suite map: RUNNING...
Test suite map: PASS
Test suite logged to:
/code/mess/cabal-type-class-default-impl-rebuild/dist-newstyle/build/x86_64-linux/ghc-8.6.5/demo-0.1.0.0/t/map/test/demo-0.1.0.0-map.log
1 of 1 test suites (1 of 1 test cases) passed.
```

**Comment by phadej** [comment]
But seems I can reproduce with GHC-8.4.4; so maybe GHC-8.4.x `--make` was broken; as `GHC` doesn't recompile `MyInstance.hs`. I don't think `cabal` can do anything here.

(note how only `MyClass` is recompiled)

```
% cabal v2-test -w ghc-8.4.4
Build profile: -w ghc-8.4.4 -O1
In order, the following will be built (use -v for more details):
 - demo-0.1.0.0 (lib) (file src/MyClass.hs changed)
 - demo-0.1.0.0 (test:map) (dependency rebuilt)
Preprocessing library for demo-0.1.0.0..
Building library for demo-0.1.0.0..
[1 of 2] Compiling MyClass          ( src/MyClass.hs, /code/mess/cabal-type-class-default-impl-rebuild/dist-newstyle/build/x86_64-linux/ghc-8.4.4/demo-0.1.0.0/build/MyClass.o )
Preprocessing test suite 'map' for demo-0.1.0.0..
Building test suite 'map' for demo-0.1.0.0..
[1 of 1] Compiling Main             ( test/MyTest.hs, /code/mess/cabal-type-class-default-impl-rebuild/dist-newstyle/build/x86_64-linux/ghc-8.4.4/demo-0.1.0.0/t/map/build/map/map-tmp/Main.o ) [MyClass changed]
Linking /code/mess/cabal-type-class-default-impl-rebuild/dist-newstyle/build/x86_64-linux/ghc-8.4.4/demo-0.1.0.0/t/map/build/map/map ...
Running 1 test suites...
Test suite map: RUNNING...
map: Prelude.undefined
CallStack (from HasCallStack):
  error, called at libraries/base/GHC/Err.hs:79:14 in base:GHC.Err
  undefined, called at src/MyClass.hs:4:13 in demo-0.1.0.0-inplace:MyClass
Test suite map: FAIL
Test suite logged to:
/code/mess/cabal-type-class-default-impl-rebuild/dist-newstyle/build/x86_64-linux/ghc-8.4.4/demo-0.1.0.0/t/map/test/demo-0.1.0.0-map.log
0 of 1 test suites (0 of 1 test cases) passed.
cabal: Tests failed for test:map from demo-0.1.0.0.
```

**Comment by yeputons** [comment]
Yes, looks like it was fixed in GHC 8.6.4, see [release notes](https://downloads.haskell.org/~ghc/8.6.4/docs/html/users_guide/8.6.4-notes.html):

> A bug wherein changes in default method definitions would fail to trigger recompilation has been fixed ([Trac #15970](https://gitlab.haskell.org/ghc/ghc/issues/15970)).

So, **workarounds**:

* Option 1: Update GHC to 8.6.4 or newer
* Option 2: pass `--ghc-options=-fforce-recomp` to `cabal` so GHC will recompile all files

Closing this issue as it's not about Cabal and is already fixed in a newer version of GHC.
